dice_ml/explainer_interfaces/dice_KD.py

```diff
@@ -0,0 +1,236 @@
+        return np.array([preds[(self.num_ouput_nodes-1):] for preds in temp_preds], dtype=np.float32)
+
+    def do_cf_initializations(self, total_CFs, algorithm, features_to_vary):
+        """Intializes CFs and other related variables."""
+
+        self.cf_init_weights = [total_CFs, algorithm, features_to_vary]
+
+        if algorithm == "RandomInitCF":
+            # no. of times to run the experiment with random inits for diversity
+            self.total_random_inits = total_CFs
+            self.total_CFs = 1          # size of counterfactual set
+        else:
+            self.total_random_inits = 0
+            self.total_CFs = total_CFs  # size of counterfactual set
+
+        # freeze those columns that need to be fixed
+        if features_to_vary != self.features_to_vary:
+            self.features_to_vary = features_to_vary
+            self.feat_to_vary_idxs = self.data_interface.get_indexes_of_features_to_vary(features_to_vary=features_to_vary)
+            self.freezer = tf.constant([1.0 if ix in self.feat_to_vary_idxs else 0.0 for ix in range(len(self.minx[0]))])
+
+
+        # CF initialization
+        if len(self.cfs) != self.total_CFs:
+            self.cfs = []
+            for ix in range(self.total_CFs):
+                one_init = [[]]
+                for jx in range(self.minx.shape[1]):
+                    one_init[0].append(np.random.uniform(self.minx[0][jx], self.maxx[0][jx]))
+                self.cfs.append(tf.Variable(one_init, dtype=tf.float32))
+
+    def find_counterfactuals(self, query_instance, desired_class, optimizer, learning_rate, min_iter, max_iter, project_iter, loss_diff_thres, loss_converge_maxiter, verbose, init_near_query_instance, tie_random, stopping_threshold, posthoc_sparsity_param, posthoc_sparsity_algorithm):
+        """Finds counterfactuals by gradient-descent."""
+
+        # Prepares user defined query_instance for DiCE.
+        query_instance_ = self.data_interface.prepare_query_instance(query_instance=query_instance, encode=True)
+        query_instance_ = np.array([query_instance_.iloc[0].values])
+
+        # find the predicted value of query_instance
+        test_pred = self.predict_fn(tf.constant(query_instance_, dtype=tf.float32))[0][0]
+
+        if desired_class == "opposite":
+            desired_class = 1.0 - np.round(test_pred)
+        else:
+            desired_class = np.round(test_pred)
+        query_instance_copy = query_instance.copy()
+
+        #preparing query instance for conversion to pandas dataframe
+        for q in query_instance_copy:
+            query_instance_copy[q] = [query_instance_copy[q]]
+        query_instance_df = pd.DataFrame.from_dict(query_instance_copy)
+
+        start_time = timeit.default_timer()
+
+        #Making the one-hot-encoded version of query instance match the one-hot encoded version of the dataset
+        query_instance_df_dummies = pd.get_dummies(query_instance_df)
```

Comment:
why is line 84 needed? this also makes think that the one-hot encoding code here and in DiCE should be moved to a separate function, so that other functions can call it easily.  It could be a static function in explainer_base.